Repeatable RNG

sidnarayanan opened the pull request:
Class that generates a fixed number of repeatable random numbers. Various simple PDFs are implemented. Many uniform->P(x) transformations are implemented best by various sampling techniques, which kind of defeats the point of this, so I only implemented Poisson in this fashion. 

Here's a little test suite:

```C++
#include "PandaTree/Utils/interface/RRNG.h"

#include <iostream>

using namespace std;

int
main(int argc, char const* argv[])
{
  RRNG r(1234, 20);

  r.generate();
  int idx = 0;
  cout << "U[0] = " << r.uniform(idx) << endl;
  cout << "U[0] = " << r.uniform() << endl;
  cout << "U[1] = " << r.uniform() << endl;
  cout << "N[2] = " << r.normal() << endl;
  double x,y; r.normal(x,y);
  cout << "N[4,5] = " << x << " "  << y << endl;
  cout << "L[5] = " << r.landau(1, 1) << endl;

  idx = 6;
  unsigned N = r.poisson(5, idx);
  cout << "Poisson(5) = " << N << endl;
  cout << "Took " << idx - 6 << " calls" << endl;

  cout << "Poisson(15) = " << r.poisson(15) << endl;
  cout << "Poisson(15) = " << r.poisson(15) << endl;
  cout << "Poisson(15) = " << r.poisson(15) << endl;
  cout << "Poisson(15) = " << r.poisson(15) << endl;

  idx = 100;
  r.uniform(idx);

  return 0;
}
```

which prints (with `RRNG_DEBUG` on):
```
U[0] = 0.357817
U[0] = 0.357817
U[1] = 0.0129075
N[2] = -1.1717
N[4,5] = -0.425174 0.761835
L[5] = 4.7379
Poisson(5) = 6
Took 6 calls
Poisson(15) = 9
RRNG::_requestNumber forced to re-generate numbers because we ran out!
Poisson(15) = 21
RRNG::_requestNumber forced to re-generate numbers because we ran out!
Poisson(15) = 19
RRNG::_requestNumber forced to re-generate numbers because we ran out!
Poisson(15) = 17
terminate called after throwing an instance of 'std::out_of_range'
  what():  RRNG::_requestNumber accessing RN 100/20
Aborted
```

Commit:
repeatable rng implementation

Commit:
no debug

Commit:
allocate less

mitsidekick commented:
Tests passed! Plots here:

- http://t3serv001.mit.edu/~jenkins/relval/?d=PandaPhysics/PandaTree/PR-109/1

sidnarayanan commented:
It's a little gross, but does anyone object to if I put this in Framework instead of Utils? Utils depends on Objects right now, so I can't have the reverse dependency.

Commit:
getEntry seeds rng

Commit:
typo

Commit:
protect copy const

dabercro commented:
One more thing I would add is a check that someone doesn't accidentally fetch the same index twice. I had that in what I was messing around with: https://github.com/dabercro/PandaTree/blob/trying-panda-stats/Framework/src/RandomFactory.cc#L19-L22 (but I ultimately fell down the generate.py rabbit hole)

sidnarayanan commented:
Yeah, this was on my to-do list. I want to support both "give me random number N" and "give me the next random number", so in principle you could end up doing something weird like accessing indices 0,1,10, and then request the "next" number. I'll think about a good way to implement this. 

One other thing: I realized that setting the seed on the first event can make things tricky for syncing. If (as is the current use case) you need to request random numbers in a loop of variable length, a single out-of-sync event can throw subsequent events out of sync. So, I've just implemented re-seeding for every event.

dabercro commented:
> I realized that setting the seed on the first event can make things tricky for syncing. If (as is the current use case) you need to request random numbers in a loop of variable length, a single out-of-sync event can throw subsequent events out of sync. So, I've just implemented re-seeding for every event.

Yeah, I realized the same thing if we have a bunch of muons or jets to smear. The number you generate is configurable, right? We can also just make that obnoxiously large.

Commit:
protect against double-access

sidnarayanan commented:
Yeah, but you can always hit a weird event that goes over that limit. And it's not just about the number of times you sample the uniform distribution. For example, if you call `RRNG::poisson`, it basically simulates a poisson process, so it can scale arbitrarily high. Re-seeding for each event also avoids problems with TChains, where the "first event" is no longer uniquely defined.

Commit:
oops

mitsidekick commented:
Tests passed! Plots here:

- http://t3serv001.mit.edu/~jenkins/relval/?d=PandaPhysics/PandaTree/PR-109/2

sidnarayanan commented:
If there are no objections, I'll merge this this weekend.

dabercro commented on Objects/interface/EventBase.h:
```diff
@@ -71,6 +72,12 @@ namespace panda {
     HLTObjectStore triggerObjects = HLTObjectStore("triggerObjects");
 
+    //! Repeatable random number generator
+    RRNG* rng{0};
```
I'm not sure there is a good reason to make this a pointer. Generally, that's only a good idea if absence of an object is valid or you need to dynamically pick a reference. I would personally rather have an object that is always valid that generates 0 random numbers by default.

sidnarayanan replied:
My original reasoning was to avoid constructing `RRNG` if people don't want to use it. It sounds like only a fraction of the users care about it, and I'd like to avoid adding overhead for everyone else.

dabercro replied:
The overhead is tiny, and if some new user tries to access a method through a null pointer, they'll just get a very unhelpful seg fault message.

sidnarayanan replied:
I'll let @yiiyama decide if this should always be instantiated or not. I think he's the only maintainer who won't use this feature.

yiiyama replied:
I think either is fine - on one hand we do instantiate things that not everyone always use (HLTObjects etc.), but on the other hand if the user has to call `initRNG()` anyway to start using this interface, we might as well instantiate it there.

dabercro commented on Framework/src/RRNG.cc:
```diff
@@ -0,0 +1,131 @@
+  _maxEntropy(other._maxEntropy)
+{
+  _x = new double[_maxEntropy];
+  _available = new bool[_maxEntropy];
```
You should make these vectors, not variable pointers. Then the default copy (and move) constructor, which I'm not even sure when we would use, will definitely work properly.

Also, in that case, you can just resize when needing to generate a second time in an event.

dabercro replied:
The resize shouldn't make any overhead when you go back down. (The vector will still own the extra space, I think.)

dabercro replied:
We can also remake `_x, _available` into a vector of `<double, bool>` pairs, so you can iterate over them together with a fancy `iterator` for both initialization and fetching.

sidnarayanan replied:
Wait, I definitely don't think vector resizing is the right thing to do. You can easily use an arbitrary amount of memory by doing big-N calls to `RRNG::poisson`

sidnarayanan replied:
And I guess that's something we ought to discuss... do we want the copy and move constructors (forgot to implement move) to create a new copy with:
- exactly the same random numbers currently as the original
- the same seed/rng/settings, but generate a fresh set of random numbers

I picked the latter option for copy. I suppose the implicitly-defined move constructor is actually going to have the former behavior. And I implemented a copy constructor so that it could be used in `EventBase`'s copy constructor, which I believe is a pretty common use case.

dabercro replied:
Ah true. I forgot about the `EventBase`. Resizing may not be necessary either. But if the settings are the same, why make the copy generate the same numbers again?

sidnarayanan replied:
I think resizing is dangerous. 

Copy is not generating the same numbers again - it's generating `_maxEntropy` numbers from a copy of the `TRandom3` instance, which copies the state of the original rng:

```root [0] TRandom3 a(123);
root [1] a.Rndm()
(Double_t) 6.964692e-01
root [2] TRandom3 b(a);
root [3] b.Rndm()
(Double_t) 7.129553e-01
root [4] a.Rndm()
(Double_t) 7.129553e-01
```

So the behavior would be something like:
```C++
RRNG a(2);     // say it generates {0.4, 0.5}
RRNG b(a);     // say b generates {0.6, 0.7}
a.generate();   // now a contains {0.6, 0.7} too
```

Now that I think about it, maybe we do want `b` to have the same state as `a` on copy... In that case, I think I'll switch to vectors just to take advantage of the of easy copies. I still don't think dynamic resizing is a good idea though. Safer to just re-use the memory.

dabercro replied:
Yeah, I think I agree now on the resizing bit.

dabercro commented on Framework/src/RRNG.cc:
```diff
@@ -0,0 +1,131 @@
+
+void RRNG::generate() 
+{
+  _rng.RndmArray(_maxEntropy, _x);
```
Oh, and if we go the route of resizeable vectors, we can put a check here for size. If it's zero, set it to some small default number, like 10, before generating. That way, the code doesn't unhelpfully seg fault if someone forgets to call (or doesn't feel like calling) `initRNG()`

sidnarayanan replied:
I'm confused. `initRNG` creates the object. Regardless of whether there's a size check here, there's going to be a seg fault if you try to access `event.rng` without `event.initRNG`. And the constructor for `RRNG` requires a size. I suppose I can throw an error if the constructor is passed 0, but that really seems like we're babysitting the user...

dabercro replied:
I realize the resizing would be better in the access method.

We shouldn't require our level of knowledge of the EventBase header just to use this. If we're assuming the use case is that small, there's no point in making a central generator.

sidnarayanan replied:
I'm not sure what you're suggesting here, and what it has to do with the generate function. I don't think a user needs detailed knowledge of the headers to know that initializing an array of random numbers of 0 size is going to produce 0 random numbers.

dabercro replied:
Yeah, I think this is the wrong line to make the suggestion. Let me find better lines.

dabercro commented on Objects/interface/EventBase.h:
```diff
@@ -71,6 +72,12 @@ namespace panda {
+    RRNG* rng{0};
+
+    //! Initialize the random number generator
+    void initRNG(int N = 50) { delete rng; rng = new RRNG(0, N); }
```
One big reason I would want default initialization (of size 0) is to prevent the interface of the `RRNG` from leaking into `EventBase`.

dabercro commented on Objects/src/EventBase.cc:
```diff
@@ -214,6 +219,9 @@ panda::EventBase::doGetEntry_(TTree& _tree)
     triggerObjects.makeMap(*run.hlt.filters);
 
+  if (rng) 
+    rng->generate(eventNumber);
```
We should make it somehow possible to turn off the resetting at every event. As you pointed out before, that might not even be the default behavior we want. If someone new uses the feature not for syncing, we should give them the most random numbers possible. In general, I also asked if there was interest in this feature in order to avoid completely unrepeatable numbers like timestamps, as well as even less correct constant numbers. I'll put a strict outline of default behaviors that I think would be nice in the main conversation.

dabercro commented on Framework/src/RRNG.cc:
```diff
@@ -0,0 +1,131 @@
+
+double RRNG::_requestNumber(int& idx) 
+{
+  int idx_;
```
This is where we should actually check for 0 size of pre-generated numbers. If size is 0, just return a new number from the `TRandom3` instance.

dabercro commented:
So I think ideal default behavior for me (so we can easily recommend this to new users) would be the following:

- Always construct the `RRNG` object, but have zero elements automatically generated. If a number is requested and no random numbers are ever pre-generated, just directly return a number from the internal `TRandom3` generator
- Set the seed to the event number on the first event, but not after (configurable). This is probably preferable for any new novice users that come along. We can construct the `RRNG` object with a reference to the `eventNumber` member so it always knows `eventNumber` if it needs it.

We can add a method for `RRNG` that looks like this (where default values will match the default construction above):

```
/*
 * \param num_generate is the number of random numbers to pre-generate
 * \param seed is the seed to set the generator to. If 0, use the eventNumber
 * \param reset_on_event will cause the generator to reset its seed at every event (either seed argument or eventNumber)
 *                       recommended True for easiest for repeatability, False for most random numbers
 */
void settings (int num_generate = 0, int seed = 0, bool reset_on_event = False)
```
I would prefer this to placing an extra method into `EventBase`

dabercro commented:
I would be open to changing my mind about the default `reset_on_event`, but I'm just a littie concerned about the occasional repeat of event numbers.

dabercro commented:
Also, probably throw up a big warning if someone sets both `seed` and `reset_on_event`

Commit:
different options for setting seed

sidnarayanan commented:
I had a slightly different interpretation of the ideal behavior. Here's what I implemented:

- `EventBase` always constructs `RRNG`, but with default size 0. The user has to call `RRNG::setSize` to specify how many numbers should be generated. If a user requests a number and the size is 0, then they get an exception. I don't think it should silently generate a random number, since that can hide a real problem (user thinks they're generating a fixed number but aren't). 

- We talked about this before, but I think for syncing purposes, it's much better to set the seed for each event. Event numbers repeat rarely, to the point where (unless you're concerned about cryptography), it's good enough. What happens if events are skipped differently between two people? What if you have `EventBase` read a `TChain`? Should it still read the first event of the chain or the first event of each tree? Etc.

- Instead of a configuration (which requires `RRNG` to have some notion of "event" and "tree", whereas I'd prefer it be standalone), I provide `setFixedSeed` and `setSeedAddress` for the user. So the user could do:

```C++

RRNG rng{10, 5, nullptr};                // fixed seed of 5, no address set
rng.generate();                          // will use fixed seed
rng.setSeedAddress(&event.eventNumber);  // every time generate() is called, if eventNumber has changed, it resets the seed
rng.generate();                          // using addressed seed
event.getEntry(5); rng.generate();       // new seed
rng.setFixedSeed(event.eventNumber);     // now fixing the seed at the current eventNumber

```

So all three behaviors are possible (fixed user-specified seed, fixed first-event seed, every-event seed). The default behavior for `EventBase::rng` is the last option. I think this is the most flexible route.

Commit:
move was a copy of copy, oops

dabercro commented:
The only thing missing that I would like is at least making it *possible* to purposefully generate numbers on the fly without something hacky and indirect like setting a low vector size.

dabercro commented:
I mean, I can just continue to use TRandom3 directly, but then I would have to rewrite more of my code every time I want to sync.

mitsidekick commented:
Tests passed! Plots here:

- http://t3serv001.mit.edu/~jenkins/relval/?d=PandaPhysics/PandaTree/PR-109/3

Commit:
set streaming mode

sidnarayanan commented:
Okay, to close the loop here:

- Added `setStreaming(bool b = true)` to turn on streaming mode. This sets `_maxEntropy` to 0 and short-circuits the logic to directly call `TRandom3::Rndm` for any call. 
- Note that `generate()` still does one thing even if streaming mode is on: if we are using an addressed seed, and the seed has changed, then it resets the seed passed to the RNG.

If everyone's happy, let's go ahead and merge this :)

mitsidekick commented:
Tests passed! Plots here:

- http://t3serv001.mit.edu/~jenkins/relval/?d=PandaPhysics/PandaTree/PR-109/4

dabercro commented:
Happy!

Commit:
remove debug
